samples: i2s_codec: Structural fixes

VitekST opened the pull request:
During testing and further enablement on i.MX RT platforms, I devised some generic, "structural" fixes of the `i2s_codec` demo. These mainly reflect:
- duplicated Kconfig settings across targets
- invalid buffer size and discontinuous playback of the test sine wave when the sample rate is reduced to a half or quarter of the default sample rate (relevant on platforms with insufficient amount of memory)
- the audio queue (slab) was put, on relevant platforms, into a cached region, interfering with DMA transfers
- I2S peripheral and the coded were both configured as masters, this both doesn't make sense (yet it worked with wm8904 and wm8906 drivers - https://github.com/zephyrproject-rtos/zephyr/pull/92774) and this requires adjusting at the level of each target (for example when the codec is intended to be a master, typical for targets with Flexcomm peripherals having signal sharing configured)
- mapping of DMIC channels 1 to 1 (https://github.com/zephyrproject-rtos/zephyr/issues/93494)

VitekST committed:
samples: i2s_codec: Move audio slab to uncached

Move mem_slab (audio queue) to an uncached region with __nocache.

On platforms with data cache (such as platforms with the Cortex M7
core), it's desirable to have the audio data in an uncached region, so
that all data is committed in the selected location before DMA transfers
are triggered.

Playback problems observed on mimxrt1170_evk/mimxrt1176/cm7 before this
was done.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
samples: i2s_codec: Parametrise peripheral and codec position

Add Kconfig symbol USE_CODEC_CLOCK. Modify configuration for supported
boards.

This change enables to adjust the position of both the codec and I2S
peripheral on the I2S bus.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
samples: i2s_codec: Reorganise Kconfig fragments

Reorganise Kconfig fragments (prj.conf, boards/*.conf). Rename
mimxrt1170_evk@B/mimxrt1176/cm7 fragments to
mimxrt1170_evk/mimxrt1176/cm7. Change Kconfig defaults for SAMPLE_RATE
and I2S_INIT_BUFFER.

This change is meant to reduce the length of target-specific
configuration, as some of the settings are duplicated across targets.
Also, with the inclusion of the sine wave vector, the default sample
rate of this sample changed to 16 kHz.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
samples: i2s_codec: Fix buffer size

Use BLOCK_SIZE instead of the test 16 kHz sine wave length. Add a build
check so that BLOCK_SIZE doesn't overflow the test sine wave buffer.

This accounts for situations where the sample queue can't be big enough,
mainly when it doesn't fit in the device's SRAM and the count of samples
queued will be less. As the test sine wave vector contains ~42 periods,
it can be truncated.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
samples: i2s_codec: Add mimxrt595_evk DTO lic. hdr.

Add license header to the mimxrt595_evk/mimxrt595s/cm33 i2s_codec's DT
overlay.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST commented:
@anangl Anything to discuss here?

VitekST committed:
samples: i2s_codec: Modify DMIC channel mapping

Modify DMIC channel mapping (req_chan_map_lo) to account for physical
1:1 mapping between logical channel numbers and physical channel space.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

anangl commented on samples/drivers/i2s/i2s_codec/Kconfig:
```diff
@@ -17,6 +17,14 @@ config SAMPLE_FREQ
 	help
 	  Sample frequency of the system.
 
+config CODEC_MASTER
+	bool "Configure the selected codec device as a master device."
+	default true
+
+config I2S_MASTER
```
It's a configuration error to enable both these options or none of them, right? Why not have just one option that specifies which clock should be used? Named for example `USE_CODEC_CLOCK` (also to avoid using terms "master" and "slave"); not enabled by default, so only when needed `CONFIG_USE_CODEC_CLOCK=y` would be added in overlays.

anangl commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -32,7 +32,9 @@
 #define BLOCK_COUNT (INITIAL_BLOCKS + 32)
-K_MEM_SLAB_DEFINE_STATIC(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
+
+__nocache struct k_mem_slab mem_slab;
+char __nocache __aligned(4) mem_slab_buf[BLOCK_SIZE * BLOCK_COUNT];
```
Both should be `static`.

anangl commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -193,7 +193,6 @@ int main(void)
 #else
 				/* If not using DMIC, play a sine wave 440Hz */
 				mem_block = (void *)&__16kHz16bit_stereo_sine_pcm;
-				block_size = __16kHz16bit_stereo_sine_pcm_len;
```
Perhaps a build assertion ensuring that `block_size <= __16kHz16bit_stereo_sine_pcm_len` would be helpful here.

anangl commented on samples/drivers/i2s/i2s_codec/Kconfig:
```diff
@@ -5,18 +5,27 @@ source "Kconfig.zephyr"
 
+config USE_CODEC_CLOCK
+	bool "I2S BCK is generated by a selected codec device"
+	default n
```
This is redundant.
```suggestion
```

anangl commented on samples/drivers/i2s/i2s_codec/src/main.c:
```diff
@@ -62,6 +65,15 @@ static bool trigger_command(const struct device *i2s_dev_codec, enum i2s_trigger
 
 int main(void)
 {
+#ifndef CONFIG_USE_DMIC
+	BUILD_ASSERT(
+		BLOCK_SIZE <= __16kHz16bit_stereo_sine_pcm_len,
+		"BLOCK_SIZE is bigger than test sine wave buffer size."
+	);
+#endif
```
This is a bit out of context here. Why not have it next to the use of `__16kHz16bit_stereo_sine_pcm`, so basically right where `__16kHz16bit_stereo_sine_pcm_len` was previously assigned to `block_size`? I think it would be clearer then why this condition is checked.

VitekST replied:
@anangl Fine. It's where you suggested.

VitekST committed:
samples: i2s_codec: Factorise initial loop count

Replace hard-wired constant "2" with CONFIG_I2S_INIT_BUFFERS in the loop
where audio data is read and written.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>
